p2p/state.py

```diff
@@ -161,32 +186,36 @@ def msg_queue_maxsize(self) -> int:
         # run in the background notice the cancel token has been triggered and return.
         await asyncio.sleep(0)
 
-    async def request_nodes(self, node_keys: List[bytes]) -> None:
-        batches = list(partition_all(eth.MAX_STATE_FETCH, node_keys))
-        for batch in batches:
-            peer = await self.get_idle_peer()
-            now = time.time()
-            for node_key in batch:
-                self._pending_nodes[node_key] = now
+    async def request_nodes(self, node_keys: Iterable[Hash32]) -> None:
+        not_yet_requested = set(node_keys)
+        while not_yet_requested:
+            try:
+                peer = await self.get_peer_for_request(not_yet_requested)
+            except NoEligiblePeers:
+                self.logger.debug(
+                    "No idle peers have any of the trie nodes we want, sleeping a bit")
+                await self.wait(asyncio.sleep(0.2))
+                continue
+
+            candidates = list(not_yet_requested.difference(self._peer_missing_nodes[peer]))
+            batch = candidates[:eth.MAX_STATE_FETCH]
+            not_yet_requested = not_yet_requested.difference(batch)
+            self._requested_nodes[peer] = (time.time(), batch)
             self.logger.debug("Requesting %d trie nodes to %s", len(batch), peer)
             peer.sub_proto.send_get_node_data(batch)
-            self._peers_with_pending_requests[peer] = now
 
     async def _periodically_retry_timedout(self) -> None:
         while self.is_running:
             now = time.time()
-            # First, update our list of peers with pending requests by removing those for which a
-            # request timed out. This loop mutates the dict, so we iterate on a copy of it.
-            for peer, last_req_time in list(self._peers_with_pending_requests.items()):
-                if now - last_req_time > self._reply_timeout:
-                    self._peers_with_pending_requests.pop(peer)
-
-            # Now re-send requests for nodes that timed out.
             oldest_request_time = now
             timed_out = []
-            for node_key, req_time in self._pending_nodes.items():
+            # Iterate over a copy of our dict's items as we're going to mutate it.
+            for peer, (req_time, node_keys) in list(self._requested_nodes.items()):
                 if now - req_time > self._reply_timeout:
-                    timed_out.append(node_key)
+                    self.logger.debug(
+                        "Timed out waiting for %d nodes from %s", len(node_keys), peer)
+                    timed_out.extend(node_keys)
+                    self._requested_nodes.pop(peer)
```

Comment:
Can we re-structure this in a non-mutative manner.

```python
timed_out = cytoolz.valfilter(lambda v: time.time() - v[0] > self._reply_timeout, self.requested_nodes)
self._requested_nodes = cytoolz.dissoc(self._request_nodes, *timed_out.keys())
timed_out_nodes = cytoolz.concat(node_keys for _, node_keys in timed_out.values())
```

Reply:
I'd rather avoid doing that now, as it's tricky to exercise this code path and the non-mutative approach is not as clear as the current one. I'm sure I'd end up introducing bugs and we wouldn't uncover them easily. If only we had proper unit tests for those things... I really should take the time to write those

Reply:
Maybe this would be easier to test if we isolated this functionality into something stand-alone

```
class StateSyncTracker:
   ...
```

It could house all of the statefullness of state syncing which would allow us to test it's APIs without needing to spin up the full state syncer?

Reply:
(let me know if it's unclear what I mean by this and I can try to flesh out an example a bit further)

Reply:
Yeah, it'll definitely need some refactoring in order to be unit tested

Reply:
I can take a stab at it, but I'm done for the day and will be off tomorrow, so this will have to wait until next week. I'll at least try to fix the issue above with the infinite loop when no peers have the nodes we want